gh-123321: Make Parser/myreadline.c locking safe in free-threaded build

colesbury opened the pull request:
Use a `PyMutex` to avoid the race in mutex initialization. Use relaxed atomics to avoid the data race on reading _PyOS_ReadlineTState when checking for re-entrant calls.

<!-- gh-issue-number: gh-123321 -->
* Issue: gh-123321
<!-- /gh-issue-number -->

colesbury committed:
gh-123321: Make Parser/myreadline.c locking safe in free-threaded build

Use a `PyMutex` to avoid the race in mutex initialization. Use relaxed
atomics to avoid the data race on reading _PyOS_ReadlineTState when
checking for re-entrant calls.

bharel commented on Parser/myreadline.c:
```diff
@@ -373,34 +373,21 @@ PyOS_Readline(FILE *sys_stdin, FILE *sys_stdout, const char *prompt)
-        }
-    }
-
-    Py_BEGIN_ALLOW_THREADS
```
What about the GIL though? We don't release it for the regular build?

colesbury replied:
Oh, that's right. We don't need it for the `PyMutex`, which handles it internally when it blocks, but we probably still need to release it for the readline call.

bharel replied:
Does PyMutex take the GIL if it's called outside of the GIL?
Or is it supposed to be called only within the GIL?

Not next to the code atm.

colesbury replied:
You can call `PyMutex_Lock` (and `PyMutex_Unlock`) with or without the GIL.

If the GIL is currently held AND `PyMutex_Lock()` needs to wait on the mutex, it will release the GIL while it waits on the mutex and reacquire the GIL before returning.

If the GIL is not currently held, then it's not touched.

vstinner commented on Parser/myreadline.c:
```diff
@@ -373,34 +373,21 @@ PyOS_Readline(FILE *sys_stdin, FILE *sys_stdout, const char *prompt)
     }
 
-    if (_PyOS_ReadlineLock == NULL) {
-        _PyOS_ReadlineLock = PyThread_allocate_lock();
```
I understand that this PR removes a race condition on this line by using `static PyMutex _PyOS_ReadlineLock;` which is initialized statically and so avoids the race condition.

vstinner reviewed (approved):
LGTM. `static PyMutex` is convenient!

colesbury committed:
Restore removed `Py_BEGIN_ALLOW_THREADS` macro call.

The PyOS_ReadlineFunctionPointer and PyOS_StdioReadline expect to be
called with the GIL released.

bharel reviewed (approved):
LGTM, thanks for the explanations 😄
